CMakeLists.txt

```diff
@@ -0,0 +1,23 @@
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+
+find_package(Git REQUIRED QUIET)
+
+project(CPM LANGUAGES NONE)
+
+if(GIT_EXECUTABLE)
+    # Generate a git-describe version string from Git repository tags
+    execute_process(
+            COMMAND ${GIT_EXECUTABLE} tag --points-at HEAD
+            WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+            OUTPUT_VARIABLE CURRENT_CPM_VERSION
+            RESULT_VARIABLE GIT_DESCRIBE_ERROR_CODE
+            OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+    if(NOT GIT_DESCRIBE_ERROR_CODE AND NOT CURRENT_CPM_VERSION STREQUAL "")
+        string(SUBSTRING ${CURRENT_CPM_VERSION} 1 -1 CURRENT_CPM_VERSION)
+    else()
+        set(CURRENT_CPM_VERSION 1.0.0-development-version)
+    endif()
+endif()
```

Comment:
If I understand this correctly, it is to get the CPM version from the git tags currently used. I would rather prefer we make the version on the cmake be the main version source. This would require the maintainers to agree to such change of workflow.

It might be possible to link/create a github action to read the cmake version and push tags automatically 🤔

Reply:
Yes it is to extract the version from the tag we are asking. I agree it would be better to include the right version to the project but the drawback is that the user could ask to use the main branch without any tags and would receive the version that is refering to the latest tags/version but still be upstream if the have pull changes before releasing a new version... Using git to acces the tag if we are in an untaged than we will receve the warning that we are using a unversionned cpm

Reply:
I can't quite follow this. The user can get a specific version from the tags, and for debugging/testing they'll get the `main` branch. What issues are we encountering in the latter? Is it that versions are coded in the cmake file? If that's the case, more appropriately it would be to switch the workflow to a cmake project versioning and/or add release branches to remove the warnings from which we specify the release tags.

Reply:
Imagine your master branch have some commit not part of a tag, and not yet ready for a update version you will be between for example the latest version let say 0.39.0 and the next wonderful version 0.40.0 Which version to put ? 0.39.0 is impossible because you are not anymore in 0.39.0 but not yet on 0.40.0

Reply:
it could be possible to have an empty version for not released commit I agree but this required more change on the workflow

Reply:
The general approach is to append the short hash or branch to the last known version (as [`PROJECT_VERSION_TWEAK`](https://cmake.org/cmake/help/latest/variable/PROJECT_VERSION_TWEAK.html#variable:PROJECT_VERSION_TWEAK)) which would be derived from the hardcoded value in `CMakeLists.txt`. 

But, since this would not be intended for production I think it is ok to have some ambiguity in the version. It would be possible to get the latest version everytime as well by setting it to github latest release tar ball.

If we still want to get the version being requested without git, I would investigate if FetchContent exposes any variable related to how it is being called and take the value from there.